addons/project/models/project_project_stage.py

```diff
@@ -41,6 +42,16 @@ def unlink_wizard(self, stage_view=False):
         }
 
     def write(self, vals):
+        if vals.get('company_id'):
+            company = self.env['res.company'].browse(vals.get('company_id'))
+            # Checking if there is a project with a different company_id than the target one. If so raise an error since this is not allowed
+            project = self.env['project.project'].search(['&', ('stage_id', 'in', self.ids), ('company_id', '!=', vals.get('company_id'))])
+
+            if project:
+                raise exceptions.UserError(
+                    _("You are trying to change a stage's company to %s, however the stage contains a project belonging to %s. "
+                    "Please make sure that the stage only contains projects belonging to the target company.", company.name, project.company_id.name))
```

Comment:
There could be projects from multiple companies in the search, which will create a traceback when attempting to get `project.company_id.name`. However, I think you can limit the search to one project and only take that one into account, which has the added benefit of slightly better performance:
```suggestion
            project = self.env['project.project'].search(['&', ('stage_id', 'in', self.ids), ('company_id', '!=', vals.get('company_id'))], limit=1)

            if project:
                raise exceptions.UserError(
                    _("You are trying to change a stage's company to %s, however the stage contains a project belonging to %s. "
                    "Please make sure that the stage only contains projects belonging to the target company.", company.name, project.company_id.name))
```